{
 "cells": [
  {
   "cell_type": "markdown",
   "source": [
    "# Load the packages"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "fc45aec8658d2ae8"
  },
  {
   "cell_type": "code",
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import scanpy as sc"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2024-07-24T22:44:37.442383Z",
     "start_time": "2024-07-24T22:44:35.102277Z"
    }
   },
   "id": "15ca49ec0f7b229f",
   "execution_count": 1
  },
  {
   "cell_type": "markdown",
   "source": [
    "# Load the data"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "e8ef0727513dba2f"
  },
  {
   "cell_type": "code",
   "outputs": [],
   "source": [
    "adata = sc.read_h5ad('/Users/junjietang/Desktop/GeneLLM/Heimdall-main/data/adata_C57BL6J-638850.01.h5ad')"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2024-07-24T22:44:37.528983Z",
     "start_time": "2024-07-24T22:44:37.443706Z"
    }
   },
   "id": "8c5c8ff371ac38c4",
   "execution_count": 2
  },
  {
   "cell_type": "markdown",
   "source": [
    "### This dataset is a MERFISH-RNA dataset, which contains the gene expression and spatial information of the cells (23389 cells and 550 genes). The CCC inference task for this dataset is to infer whether two spatially neighbor cells communicate through specific Ligand-creceptor pairs based on the expression level of these two cells."
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "482c3720bf1d334f"
  },
  {
   "cell_type": "code",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AnnData object with n_obs × n_vars = 23389 × 550\n",
      "    obs: 'cells', 'Tissue', 'Technology', 'xcoord', 'ycoord', 'donor_label', 'donor_genotype', 'donor_sex', 'neurotransmitter', 'class', 'subclass', 'supertype', 'cluster'\n",
      "    var: 'genes'\n",
      "    obsm: 'spatial'\n",
      "    obsp: 'Agt -> Agtr1a', 'Calcb -> Calcr', 'Gal -> Galr1', 'Grp -> Grpr', 'Hgf -> Met', 'Kitl -> Kit', 'Lama1 -> Sv2b', 'Lama1 -> Sv2c', 'Pdyn -> Oprd1', 'Pdyn -> Oprk1', 'Penk -> Oprd1', 'Penk -> Oprk1', 'Ptprm -> Ptprm', 'Tac2 -> Tacr3'\n"
     ]
    }
   ],
   "source": [
    "print(adata)"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2024-07-24T22:44:37.531878Z",
     "start_time": "2024-07-24T22:44:37.529781Z"
    }
   },
   "id": "6210cc2444c7b77f",
   "execution_count": 3
  },
  {
   "cell_type": "markdown",
   "source": [
    "### The classification labels, which represent whether two spatially neighbor cells communicate through a specific Ligand-receptor pair, are list in a dict formed \"adata.obsp\". Specifically, keys represent the Ligand-receptor pair (e.g., Agt(Ligand) -> Agtr1a(Receptor)), and values represent the induced silver standard cell-to-cell interaction relationship through the ligand-receptor pair."
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "d4b9016e45216268"
  },
  {
   "cell_type": "code",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "KeysView(PairwiseArrays with keys: Agt -> Agtr1a, Calcb -> Calcr, Gal -> Galr1, Grp -> Grpr, Hgf -> Met, Kitl -> Kit, Lama1 -> Sv2b, Lama1 -> Sv2c, Pdyn -> Oprd1, Pdyn -> Oprk1, Penk -> Oprd1, Penk -> Oprk1, Ptprm -> Ptprm, Tac2 -> Tacr3)\n",
      "The number of used Ligand-receptor pairs in this dataset: 14\n"
     ]
    }
   ],
   "source": [
    "print(adata.obsp.keys())\n",
    "print(\"The number of used Ligand-receptor pairs in this dataset: \" + str(len(adata.obsp.keys())))"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2024-07-24T22:44:37.535261Z",
     "start_time": "2024-07-24T22:44:37.533241Z"
    }
   },
   "id": "9fcd3bc709fa6644",
   "execution_count": 4
  },
  {
   "cell_type": "markdown",
   "source": [
    "### The value of a specific key (i.e., Ligand-receptor pair) is cell-by-cell matrix, in which elements \"1\" represent two spatially neighbor cells are interacted through a specific Ligand-receptor pair, elements \"-1\" represent two cells are not interacted through a specific Ligand-receptor pair. The elements \"0\" are not considered in the task.\n"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "239d4049bb491e9f"
  },
  {
   "cell_type": "code",
   "outputs": [
    {
     "data": {
      "text/plain": "<23389x23389 sparse matrix of type '<class 'numpy.int32'>'\n\twith 233890 stored elements in Compressed Sparse Row format>"
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obsp['Agt -> Agtr1a']"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2024-07-24T22:44:37.539230Z",
     "start_time": "2024-07-24T22:44:37.535990Z"
    }
   },
   "id": "cb66a0cf104b566a",
   "execution_count": 5
  },
  {
   "cell_type": "code",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The unique elements in this matrix: [-1, 0, 1]\n"
     ]
    }
   ],
   "source": [
    "print(\"The unique elements in this matrix: \" + str(np.unique(np.array(adata.obsp['Agt -> Agtr1a'].todense())).tolist()))"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2024-07-24T22:44:48.330327Z",
     "start_time": "2024-07-24T22:44:37.539923Z"
    }
   },
   "id": "d2b488e08cac916b",
   "execution_count": 6
  },
  {
   "cell_type": "markdown",
   "source": [
    "### In summary, the target of this task is as follow. Given the expression value of two cells and a specific Ligand-receptor pair, train a model with binary output to determine whether the two cells interact through such a ligand-receptor pair (i.e., label \"-1\" or \"1\")."
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "f3701c5de3e0837a"
  },
  {
   "cell_type": "code",
   "outputs": [],
   "source": [],
   "metadata": {
    "collapsed": false
   },
   "id": "22d6839b0958be65"
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
