{
 "cells": [
  {
   "cell_type": "code",
   "id": "initial_id",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "##Load the packages\n",
    "import scipy.sparse as sp\n",
    "import pickle\n",
    "import numpy as np\n",
    "import scanpy\n",
    "import pandas as pd\n",
    "from itertools import chain\n",
    "import os\n",
    "import matplotlib.pyplot as plt\n",
    "from sklearn.neighbors import NearestNeighbors"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "# The function for preprocessing the data and extracting the silver standard of Cell-cell-communcation",
   "id": "7c78c595e7aa7361"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "def Preprocess_Atlas(tissue_dict,tissue_index,\n",
    "                    cellchat_ligand_receptor_filedir,\n",
    "                    if_filtercells = False,min_genes = 200,\n",
    "                    if_filtergenes = False,min_cells = 3,\n",
    "                    if_HVG = False,N_HVG = 3000,\n",
    "                    num_spatialneighbor = 10):\n",
    "    #################################################################################################\n",
    "    print(\"tissue_index: \" + str(tissue_index + 1) + \" Tissue: \" + tissue_dict['Tissue'][tissue_index])\n",
    "    slice_name = tissue_dict['Tissue'][tissue_index]\n",
    "    #################################################################################################\n",
    "    \n",
    "    ##1. Load the spatial transcriptomics data\n",
    "    #################################################################################################\n",
    "    if tissue_dict['Technology'][tissue_index] == '10XVisium' or tissue_dict['Technology'][\n",
    "        tissue_index] == '10XVisiumHD':\n",
    "        ##Gene expression\n",
    "        expression_data = tissue_dict['Gene_expression'][tissue_index].X\n",
    "        ##Cell label and gene label\n",
    "        genes = (tissue_dict['Gene_expression'][tissue_index].var_names).tolist()\n",
    "        cells = (tissue_dict['Gene_expression'][tissue_index].obs_names).tolist()\n",
    "    if tissue_dict['Technology'][tissue_index] == 'MERFISH_Zeng' or tissue_dict['Technology'][\n",
    "        tissue_index] == 'MERFISH_Zhuang':\n",
    "        ##Gene expression\n",
    "        expression_data = tissue_dict['Gene_expression'][tissue_index]['gene_expression']\n",
    "        ##Cell label and gene label\n",
    "        genes = tissue_dict['Gene_expression'][tissue_index]['feature_names']['feature_names'].tolist()\n",
    "        cells = tissue_dict['Gene_expression'][tissue_index]['feature_names']['cellname'].tolist()\n",
    "    ##Build the adata\n",
    "    adata = scanpy.AnnData(X=expression_data, obs={'cells': cells}, var={'genes': genes})\n",
    "    adata.obs_names = cells\n",
    "    adata.var_names = genes\n",
    "    del expression_data, cells, genes\n",
    "    ##Add Tissue name to the adata\n",
    "    adata.obs[\"Tissue\"] = tissue_dict['Tissue'][tissue_index]\n",
    "    ##Add Technology to the adata\n",
    "    adata.obs[\"Technology\"] = tissue_dict['Technology'][tissue_index]\n",
    "    ##Add spatial location information to the adata\n",
    "    if tissue_dict['Technology'][tissue_index] == '10XVisium' or tissue_dict['Technology'][\n",
    "        tissue_index] == '10XVisiumHD':\n",
    "        x_coord_name = \"pxl_row_in_fullres\"\n",
    "        y_coord_name = \"pxl_col_in_fullres\"\n",
    "        tissue_dict['Spatial_location'][tissue_index].index = tissue_dict['Spatial_location'][tissue_index]['barcode']\n",
    "    if tissue_dict['Technology'][tissue_index] == 'MERFISH_Zeng' or tissue_dict['Technology'][\n",
    "        tissue_index] == 'MERFISH_Zhuang':\n",
    "        x_coord_name = \"x\"\n",
    "        y_coord_name = \"y\"\n",
    "    adata.obs[\"xcoord\"] = tissue_dict['Spatial_location'][tissue_index][x_coord_name][adata.obs.index]\n",
    "    adata.obs[\"ycoord\"] = tissue_dict['Spatial_location'][tissue_index][y_coord_name][adata.obs.index]\n",
    "    ##Add other meta information to the adata\n",
    "    if len((tissue_dict['Metadata'][tissue_index])) > 0:\n",
    "        adata.obs[['donor_label', 'donor_genotype', 'donor_sex', 'neurotransmitter', 'class', 'subclass', 'supertype',\n",
    "                   'cluster']] = tissue_dict['Metadata'][tissue_index][\n",
    "            ['donor_label', 'donor_genotype', 'donor_sex', 'neurotransmitter', 'class', 'subclass', 'supertype',\n",
    "             'cluster']]\n",
    "    adata.obsm['spatial'] = np.array(tissue_dict['Spatial_location'][tissue_index][[x_coord_name, y_coord_name]])\n",
    "    ##Data preprocessing\n",
    "    if if_filtercells:\n",
    "        scanpy.pp.filter_cells(adata, min_genes=min_genes)\n",
    "    if if_filtergenes:\n",
    "        scanpy.pp.filter_genes(adata, min_cells=min_cells)\n",
    "    adata_raw = adata.copy()\n",
    "    adata.X_raw = adata.X.copy()\n",
    "    scanpy.pp.normalize_total(adata, target_sum=1e4)\n",
    "    scanpy.pp.log1p(adata)\n",
    "    if if_HVG:\n",
    "        scanpy.pp.highly_variable_genes(adata, n_top_genes=N_HVG, flavor='seurat_v3')\n",
    "        adata = adata[:, adata.var.highly_variable]\n",
    "    #################################################################################################\n",
    "    \n",
    "    ##2. Load the CellChat ligand-receptor database\n",
    "    #################################################################################################\n",
    "    ##load the Ligand-Receptor pairs from CellChat\n",
    "    cellchat_ligand_receptor = pd.read_csv(cellchat_ligand_receptor_filedir, header=None)\n",
    "    cellchat_ligand_receptor = np.array(cellchat_ligand_receptor.iloc[:, 0]).tolist()\n",
    "    for i in range(len(cellchat_ligand_receptor)):\n",
    "        cellchat_ligand_receptor[i] = cellchat_ligand_receptor[i].replace(\" \", \"\")\n",
    "    Ligand_list = [cellchat_ligand_receptor[i].split(\"-\")[0] for i in range(len(cellchat_ligand_receptor))]\n",
    "    Ligand_list_unique = np.unique(np.array(Ligand_list)).tolist()\n",
    "    Receptor_list0 = [cellchat_ligand_receptor[i].split(\"-\")[1] for i in range(len(cellchat_ligand_receptor))]\n",
    "    Receptor_list = []\n",
    "    for j in range(len(Receptor_list0)):\n",
    "        if '+' in Receptor_list0[j]:\n",
    "            Receptor_list.append(Receptor_list0[j].replace(\"(\", \"\").replace(\")\", \"\").split(\"+\"))\n",
    "        else:\n",
    "            Receptor_list.append([Receptor_list0[j]])\n",
    "    LR_in_bool_list = []\n",
    "    use_Ligand_list = []\n",
    "    use_Receptor_list = []\n",
    "    for LR_index in range(len(cellchat_ligand_receptor)):\n",
    "        ligand_cur = cellchat_ligand_receptor[LR_index].split(\"-\")[0]\n",
    "        receptor_cur = cellchat_ligand_receptor[LR_index].split(\"-\")[1]\n",
    "        if \"+\" in receptor_cur:\n",
    "            receptor_cur = receptor_cur.replace(\"(\", \"\").replace(\")\", \"\").split(\"+\")\n",
    "        # LR_in_bool = (np.sum(np.isin(adata.var_names,receptor_cur)) == len(receptor_cur)) *(np.sum(np.isin(adata.var_names,ligand_cur)) == len(ligand_cur))\n",
    "        LR_in_bool = (np.sum(~np.isin(receptor_cur,adata.var_names)) == 0) * (\n",
    "                np.sum(np.isin(adata.var_names, ligand_cur)) > 0)\n",
    "        # LR_in_bool = (np.sum(np.isin(adata.var_names, receptor_cur)) == len(receptor_cur)) * (\n",
    "        #         np.sum(np.isin(adata.var_names, ligand_cur)) > 0)\n",
    "        LR_in_bool_list.append(LR_in_bool)\n",
    "        if LR_in_bool:\n",
    "            if isinstance(receptor_cur, list):\n",
    "                receptor_cur = np.array(receptor_cur)[np.where(np.isin(receptor_cur, adata.var_names))[0]].tolist()\n",
    "            use_Ligand_list.append(ligand_cur)\n",
    "            use_Receptor_list.append(receptor_cur)\n",
    "    Ligand_to_Receptor_list = [{'Ligand': use_Ligand_list[i],\n",
    "                                'Receptor': use_Receptor_list[i]} for i in range(len(use_Ligand_list))]\n",
    "    #################################################################################################\n",
    "    \n",
    "    ##3. Construct the ligand-receptor interaction matrix of the spatial transcriptomics data\n",
    "    #################################################################################################\n",
    "    cell_spatial = adata.obsm['spatial']\n",
    "    spatialnn_use = NearestNeighbors(n_neighbors=num_spatialneighbor + 1, algorithm='auto')\n",
    "    spatialnn_use.fit(cell_spatial)\n",
    "    spatialdist, spatial_neighborindex = spatialnn_use.kneighbors(cell_spatial)\n",
    "    ##\n",
    "    index_mat = np.tile(np.arange(spatial_neighborindex.shape[0])[:, np.newaxis], (1, spatial_neighborindex.shape[1]))\n",
    "    ##Random permutation\n",
    "    np.random.seed(0)\n",
    "    random_permutation_num = 100\n",
    "    # random_permutation_num = 1\n",
    "    index_mat_randomperm_list = []\n",
    "    spatial_neighborindex_randomperm_list = []\n",
    "    for i in range(random_permutation_num):\n",
    "        random_index = np.random.permutation(np.arange(cell_spatial.shape[0]))\n",
    "        cell_spatial_randomperm = cell_spatial[random_index, :]\n",
    "        spatialnn_use_randomperm = NearestNeighbors(n_neighbors=num_spatialneighbor + 1, algorithm='auto')\n",
    "        spatialnn_use_randomperm.fit(cell_spatial_randomperm)\n",
    "        spatialdist_randomperm, spatial_neighborindex_randomperm = spatialnn_use_randomperm.kneighbors(cell_spatial_randomperm)\n",
    "        ##\n",
    "        index_mat_randomperm = np.tile(np.arange(spatial_neighborindex_randomperm.shape[0])[:, np.newaxis],\n",
    "                            (1, spatial_neighborindex_randomperm.shape[1]))\n",
    "        spatial_neighborindex_randomperm_list.append(spatial_neighborindex_randomperm)\n",
    "        index_mat_randomperm_list.append(index_mat_randomperm)\n",
    "    ##\n",
    "    LRI_intensity_Cellbycell_list = []\n",
    "    for LR_index in range(len(Ligand_to_Receptor_list)):\n",
    "        # print(\"The LR pair: \" + str(LR_index))\n",
    "        Ligand_cur = Ligand_to_Receptor_list[LR_index]['Ligand']\n",
    "        Receptor_cur = Ligand_to_Receptor_list[LR_index]['Receptor']\n",
    "        Receptor_cur = np.array(Receptor_cur)[(np.isin(Receptor_cur, adata.var_names))].tolist()\n",
    "        Ligand_index_inusegene = np.where(np.isin(adata.var_names, Ligand_cur))[0]\n",
    "        Receptor_index_inusegene = np.where(np.isin(adata.var_names, Receptor_cur))[0]\n",
    "        ##\n",
    "        Ligand_expression = np.array(adata.X[:, Ligand_index_inusegene].todense())\n",
    "        Receptor_expression = np.array(np.max(adata.X[:, Receptor_index_inusegene].todense(), axis=1))\n",
    "        LRI_intensity = np.power(Ligand_expression[index_mat[:,1:].ravel(),:] *\n",
    "                                 Receptor_expression[spatial_neighborindex[:,1:].ravel()],0.5)\n",
    "        ##Calculate the null distribution\n",
    "        LRI_intensity_randperm = np.power(Ligand_expression[index_mat_randomperm_list[0][:,1:].ravel(),:] *\n",
    "                                 Receptor_expression[spatial_neighborindex_randomperm_list[0][:,1:].ravel()],0.5)\n",
    "        threshold_perm = np.quantile(LRI_intensity_randperm,0.99)\n",
    "        if np.sum(LRI_intensity) > 0:\n",
    "            LRI_intensity = LRI_intensity.ravel()\n",
    "            sender_index = index_mat[:,1:].ravel()\n",
    "            reciever_index = spatial_neighborindex[:,1:].ravel()\n",
    "            LRI_intensity = LRI_intensity\n",
    "            dict_cur = {'LRpair': Ligand_to_Receptor_list[LR_index],\n",
    "                                                  'LRI_intensity': LRI_intensity,\n",
    "                                                  \"If_significant_interaction\": LRI_intensity>threshold_perm,\n",
    "                                                  'sender_index': sender_index,\n",
    "                                                  'sender_cell': adata.obs.index[sender_index],\n",
    "                                                  'reciever_index': reciever_index,\n",
    "                                                  'reciever_cell': adata.obs.index[reciever_index]}\n",
    "            LRI_intensity_Cellbycell_list.append(dict_cur)\n",
    "            trans_value = []\n",
    "            for i in range(len(dict_cur[\"If_significant_interaction\"])):\n",
    "                if dict_cur[\"If_significant_interaction\"][i]:\n",
    "                    trans_value.append(1)\n",
    "                else:\n",
    "                    trans_value.append(-1)\n",
    "            adj_matrix_sparse = sp.csr_matrix((trans_value, (sender_index, reciever_index)),\n",
    "                                              shape=(adata.shape[0], adata.shape[0]))\n",
    "            ##Add to the adata\n",
    "            if isinstance(Ligand_to_Receptor_list[LR_index]['Receptor'],list):\n",
    "                LR_name = Ligand_to_Receptor_list[LR_index]['Ligand']+ \" -> \" + Ligand_to_Receptor_list[LR_index]['Receptor'][0] + \"+\" + Ligand_to_Receptor_list[LR_index]['Receptor'][1]\n",
    "            else:\n",
    "                LR_name = Ligand_to_Receptor_list[LR_index]['Ligand'] + \" -> \" + \\\n",
    "                          Ligand_to_Receptor_list[LR_index]['Receptor']\n",
    "            adata_raw.obsp[LR_name] = adj_matrix_sparse\n",
    "    print(\"The LR pair in data: \" + str(len(LRI_intensity_Cellbycell_list)))\n",
    "    # print(\"The number of LR edge: \"+ str(np.sum([len(LRI_intensity_Cellbycell_list[i]['LRI_intensity']) for i in range(len(LRI_intensity_Cellbycell_list))])))\n",
    "    ##Without self\n",
    "    spatial_neighborindex_withoutself = spatial_neighborindex[:, 1:]\n",
    "    index_mat_withoutself = np.tile(np.arange(spatial_neighborindex_withoutself.shape[0])[:, np.newaxis], (1, spatial_neighborindex_withoutself.shape[1]))\n",
    "    spatial_neighbor_index_withoutself = np.stack((index_mat_withoutself.ravel(),spatial_neighborindex_withoutself.ravel()))\n",
    "    print(\"#################################################################################################\")\n",
    "    #################################################################################################\n",
    "    ##4. Return the adata and the ligand-receptor interaction matrix\n",
    "    #################################################################################################\n",
    "    return adata_raw,  LRI_intensity_Cellbycell_list,slice_name\n",
    "\n",
    "##mkdir\n",
    "def mkdir(path):\n",
    "    folder = os.path.exists(path)\n",
    "    if not folder:\n",
    "        os.makedirs(path)\n",
    "    return path"
   ],
   "id": "aeab37196b9d0e88",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "# Set up the ditrectory of the loading data and the saving objects",
   "id": "ec28a7815418e5d0"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "filedir_main = \"D:/Project_genesllm/Data/\"",
   "id": "4a5b01c2ba68defb",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "# Obtain the slices information of the collected spatial transcriptomics data",
   "id": "305a9d971656a17d"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "##All the slices\n",
    "with open(filedir_main + \"tissue_dict_MouseBrain.pickle\", 'rb') as pickle_file:\n",
    "    tissue_dict = pickle.load(pickle_file)\n",
    "################################################################################################\n",
    "print(\"The summary of tissue_dict_MouseBrain:\")\n",
    "print(tissue_dict.keys())\n",
    "print(np.unique(tissue_dict['Technology']))\n",
    "num_tissue = len(tissue_dict['Tissue'])\n",
    "print(\"The number of tissues: \" + str(num_tissue))\n",
    "##Merfish slices index\n",
    "Merfish_Zeng_index = np.where(np.array(tissue_dict['Technology']) == 'MERFISH_Zeng')[0]\n",
    "Merfish_Zhuang_index = np.where(np.array(tissue_dict['Technology']) == 'MERFISH_Zhuang')[0]\n",
    "print(\"#################################################################################################\")"
   ],
   "id": "f0f79d8d8193f84b",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "# Extract the silver standard of Cell-cell-communication slices by slices",
   "id": "556a95213a6a6123"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "print(\"The MERFISH slices in Zeng:\" + str(len(Merfish_Zeng_index)))\n",
    "print(\"The MERFISH slices in Zhuang:\" + str(len(Merfish_Zhuang_index)))\n",
    "##\n",
    "tissue_index_type_count = 0\n",
    "for tissue_index_type in [Merfish_Zeng_index,Merfish_Zhuang_index]:\n",
    "    tissue_index_type_count = tissue_index_type_count + 1\n",
    "    for tissue_index_use in tissue_index_type:\n",
    "            adata_raw, LRI_intensity_Cellbycell_list, slice_name = Preprocess_Atlas(tissue_dict=tissue_dict,\n",
    "                                                                           tissue_index=tissue_index_use,\n",
    "                                                                           cellchat_ligand_receptor_filedir = filedir_main + \"Mouse_LRI_Cellchat.txt\",\n",
    "                                                                           if_filtercells=False, min_genes=200,\n",
    "                                                                           if_filtergenes=False, min_cells=3,\n",
    "                                                                           if_HVG=False, N_HVG=3000,\n",
    "                                                                           num_spatialneighbor=10)\n",
    "            if tissue_index_type_count == 1:\n",
    "                mkdir(filedir_main + \"CCC_adata_Zeng_mousebrain550/\")\n",
    "                adata_raw.write(filedir_main + \"CCC_adata_Zeng_mousebrain550/\" + slice_name + \".h5ad\")\n",
    "            else:\n",
    "                mkdir(filedir_main + \"CCC_adata_MERFISH_mousebrain1122/\")\n",
    "                adata_raw.write(filedir_main + \"CCC_adata_MERFISH_mousebrain1122/\" + slice_name + \".h5ad\")\n",
    "            ##\n",
    "            del adata_raw, LRI_intensity_Cellbycell_list, slice_name"
   ],
   "id": "ec0e073d05e64823",
   "outputs": [],
   "execution_count": null
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
